File: packages/tale_shared/src/tale_shared/utils/hashing.py
"""SHA-256 hashing utilities for content deduplication."""

import hashlib
from pathlib import Path


def compute_file_hash(file_path: str | Path) -> str:
    """Compute SHA-256 hash of a file's content.

    Args:
        file_path: Path to the file.

    Returns:
        Hex-encoded SHA-256 hash string.
    """
    sha256 = hashlib.sha256()
    with open(file_path, "rb") as f:
        while chunk := f.read(8192):
            sha256.update(chunk)
    return sha256.hexdigest()


def compute_content_hash(content: str | bytes) -> str:
    """Compute SHA-256 hash of in-memory content.

    Args:
        content: String or bytes to hash. Strings are encoded as UTF-8.

    Returns:
        Hex-encoded SHA-256 hash string.
    """
    if isinstance(content, str):
        content = content.encode("utf-8")
    return hashlib.sha256(content).hexdigest()

File: packages/shared/src/utils/hashing.ts
/**
 * SHA-256 hashing utilities for content deduplication.
 */

import { createHash } from 'node:crypto';
import { createReadStream } from 'node:fs';

/**
 * Compute the SHA-256 hash of a file's content, streaming to keep memory flat.
 *
 * Returns a hex-encoded SHA-256 hash string.
 */
export function computeFileHash(filePath: string): Promise<string> {
  return new Promise<string>((resolve, reject) => {
    const sha256 = createHash('sha256');
    const stream = createReadStream(filePath, { highWaterMark: 8192 });
    stream.on('data', (chunk) => sha256.update(chunk));
    stream.on('error', reject);
    stream.on('end', () => resolve(sha256.digest('hex')));
  });
}

/**
 * Compute the SHA-256 hash of in-memory content. Strings are encoded as UTF-8.
 *
 * Returns a hex-encoded SHA-256 hash string.
 */
export function computeContentHash(content: string | Uint8Array): string {
  const sha256 = createHash('sha256');
  sha256.update(
    typeof content === 'string' ? Buffer.from(content, 'utf-8') : content,
  );
  return sha256.digest('hex');
}
